### PR TITLE
Removes domain and port for rover service resources

### PR DESCRIPTION
**Pull request by aninternetof**
Using "localhost:5000" to access the Flask service works great on development PCs, but doesn't work in the final use-case of running a browser on one PC and the Flask service on the rover.

This pull request removes the localhost domain in the url and creates a variable for the rover's IP address. 

For now, we'll assume that the Flask service and the webpage are being hosted from the same place, and set the IP address to an empty string.  Since browsers don't like to let you access other ports than the one that served you the webpage, we'll need to access the Flask API through Lighttpd's port 80. To set up lightly to forward our API calls on to Flask's port 5000, follow these steps:

* Edit your lighttpd.conf in these two ways
  * Uncomment the line that says "mod_proxy" to enable mod_proxy
  * Add this somewhere:
   ```
#### proxy module
## read proxy.txt for more info
proxy.server               = ( "/api/v1/" =>
                               (
                                 (
                                   "host" => "127.0.0.1",
                                   "port" => 5000
                                 )
                               )
                             )
   ```

* Restart lighttpd
`sudo /etc/init.d/lighttpd restart`

### CODE REVIEW

**Comment by aninternetof** [comment]
@cabarnes , agreed. I created rover-api.js and moved anything that uses roverResource into it. It's still not in great shape, because a lot of the rover-api functions are still entangled with front-end mission-control stuff. But it's gonna be a lot more work to separate all of that, so I'll save it for another time. Plus, @wavetylor is gonna teach us about good js namespacing and bundling and whatnot at some point.

File: www/js/mission-control.js [hunk 1]
```diff
@@ -10,6 +10,9 @@ var blocklyArea = document.getElementById('blocklyArea');
 var blocklyDiv = document.getElementById('blocklyDiv');
 var eventQueue = [];
 
+var roverDomain = ''; //Later, if the rover is not hosting the webpage, its address will go here
+var roverApiPath = '/api/v1/';
```
Comment (cabarnes) [inline www/js/mission-control.js]: Shouldn't both of these lines be moved to rover-api.js?

File: www/js/mission-control.js [hunk 2]
```diff
@@ -212,7 +215,7 @@ function loadDesign(name) {
 function acceptName() {
```
Comment (cabarnes) [inline www/js/mission-control.js]: I think all of these functions (saveDesign, refreshSavedBds, loadDesign, and acceptName) all moved to rover-api.js. Shouldn't they all be removed from this file?

File: www/js/mission-control.js (hunk 1 above)
Comment (aninternetof) [inline www/js/mission-control.js]: Yup. Moved now.

File: www/js/mission-control.js (hunk 2 above)
Comment (aninternetof) [inline www/js/mission-control.js]: Whoops, yes. They are deleted now.

**Comment by aninternetof** [comment]
Note the change in the lighttpd.conf instructions at the top. `"host" => "localhost"` didn't work, but `"host => "127.0.0.1"` does work.

File: www/js/rover-api.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
+/*----- MISC GLOBALS -----*/
+var roverDomain = ''; //Later, if the rover is not hosting the webpage, its address will go here
+var roverApiPath = '/api/v1/';
+
+/*----- HELPER FUNCTIONS -----*/
+function roverResource(resource) {
+	return roverDomain+roverApiPath+resource;
```
Comment (wavetylor) [inline www/js/rover-api.js]: I think this could be a nice utility function if you extended it to accept arrays, something like
` return roverDomain + roverApiPath + (Array.isArray(resource) ? resource.join('/') : resource);`
Then line 47 could be:

`$.get(roverResource(['blockdiagrams', name]), function(response) {...}`

File: www/js/rover-api.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		else
+			designName = name;
+		$('a#downloadLink').attr("href", "saved-bds/"+designName+".xml");
+		$('a#downloadLink').attr("download", designName+".xml");
+		$('a#designNameArea').text(designName);
+
+		hideBlockByComment("MAIN EVENT HANDLER LOOP");
+		var hiddenBlock;
+		var allBlocksHidden = true;
+		for (hiddenBlock of blocksToHide) {
```
Comment (wavetylor) [inline www/js/rover-api.js]: where does blocksToHide come from?

File: www/js/rover-api.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+		if (allBlocksHidden)
+			showBlock('always');
+	}).error(function(){
+			alert("There was an error loading your design from the rover");
+	});
+	updateCode();
+}
+
+function acceptName() {
+	designName = $('input[name=designName]').val();
```
Comment (wavetylor) [inline www/js/rover-api.js]: is this intended to be a global variable?

File: www/js/rover-api.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+	$.get(roverResource('blockdiagrams'), function(json){
+		var duplicate = false;
+		for (var i=0; i<json.length; i++){
```
Comment (wavetylor) [inline www/js/rover-api.js]: assuming is in an array:
`var duplicate = json.indexOf(designName) > -1`
will get rid of the entire loop

File: www/js/rover-api.js (hunk 3 above)
Comment (aninternetof) [inline www/js/rover-api.js]: Oh, cool, yeah. I'll make that change.

File: www/js/rover-api.js (hunk 4 above)
Comment (aninternetof) [inline www/js/rover-api.js]: Will do. I waffle on that principle, but I think you're right,

File: www/js/rover-api.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+			if (json[i] == designName)
+				duplicate = true;
+		}
+
+		if (designName === ''){
```
Comment (wavetylor) [inline www/js/rover-api.js]: we can check this condition before we make the get request and save ourselves a whole round trip to the server. Also this test will fail if designName is null or undefined. In js you can abuse its type casting to succinctly test this with `if(!designName) {}`. But this will also reject the value int `0` but not `'0'`

File: www/js/rover-api.js (hunk 4 above)
Comment (aninternetof) [inline www/js/rover-api.js]: It's an array in mission-control.js. I moved these functions over from mission-control.js, but there are still lots of entanglements. At some point it will be good to go detangle them.

File: www/js/rover-api.js (hunk 5 above)
Comment (aninternetof) [inline www/js/rover-api.js]: Yeah, it is declared/initialized in mission-control.js. Another bad entanglement.

File: www/js/rover-api.js (hunk 6 above)
Comment (aninternetof) [inline www/js/rover-api.js]: Oh cool. Will change.

File: www/js/rover-api.js [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+function refreshSavedBds() {
+	$.get(roverResource('blockdiagrams'), function(json){
+		if (!json.result.length){
+			$('#savedDesignsArea').text("There are no designs saved on this rover");
+		} else {
+			$('#savedDesignsArea').empty();
+			json.result.forEach(function(entry) {
+				$('#savedDesignsArea').append("<a href='#' class='button' style='margin:10px;' onclick='return loadDesign(\""+entry+"\")'>"+entry+"</a>");
```
Comment (wavetylor) [inline www/js/rover-api.js]: I'm not a huge fan of inlining dom like this. Its hard to maintain and cumbersome to read and interpret. Jquery has a fairly nice fluid syntax for this:

```
$(document.createElement('a')).addClass('button')
  .html(entry)
  .attr('href', '#')
  .css('margin', '10px')
  .appendTo($("savedDesignsArea"))
  .click(function() {
    return loadDesign(entry);
  });
```
This is definitely just a style suggestion though!

File: www/js/rover-api.js [hunk 9]
```diff
@@ -0,0 +1,98 @@
+/*----- MISC GLOBALS -----*/
```
Comment (wavetylor) [inline www/js/rover-api.js]: I'm not sure whats going on, but these tab spacing looks super meaty

Comment (aninternetof) [inline www/js/rover-api.js]: Fixing...

File: www/js/rover-api.js (hunk 8 above)
Comment (aninternetof) [inline www/js/rover-api.js]: That looks nice. I'll give it a try!
